### PR TITLE
EIP1-8864: Adding OE rejected template preview

### PR DESCRIPTION
**Pull request by ChristianSoftwire**
## Describe your changes

Adding OE rejected template preview endpoint 

## Issue ticket number and link

https://dluhcdigital.atlassian.net/browse/EIP1-8864

## Test link

http://localhost:3002/

## Checklist before requesting a review

- [x] I double checked that ACs on the ticket are met by this code update
- [ ] I have added testing steps to the ticket
- [ ] I have used valid and semantic HTML code to structure my component/page
- [ ] I have added unit tests to new code and updated existing tests where needed
- [ ] I have checked and fixed validation/a11y issues
- [ ] The new/updated UI is responsive (mobile, tablet, desktop)
- I have tested the changes using
  - [ ] Firefox
  - [ ] Chrome/Brave/Opera
  - [ ] Edge
  - [ ] Safari

### CODE REVIEW

File: src/main/kotlin/uk/gov/dluhc/notificationsapi/config/NotifyClientConfiguration.kt [hunk 1]
```diff
@@ -215,14 +269,25 @@ class VoterCardNotifyEmailTemplateConfiguration(
...
+    val postal: PostalNotifyLetterTemplateConfiguration,
+    val proxy: ProxyNotifyLetterTemplateConfiguration,
+    val overseas: OverseasNotifyLetterTemplateConfiguration
+)
+
+abstract class AbstractNotifyLetterTemplateConfiguration(
     val sourceType: SourceType,
     val receivedEnglish: String?,
     val receivedWelsh: String?,
-    val approvedEnglish: String?,
```
Comment (TomosSlater) [inline src/main/kotlin/uk/gov/dluhc/notificationsapi/config/NotifyClientConfiguration.kt:224]: Is there a reason why the email template has "approved" still but the letter one has "rejected"?

File: src/main/kotlin/uk/gov/dluhc/notificationsapi/config/NotifyClientConfiguration.kt [hunk 2]
```diff
@@ -139,6 +182,12 @@ class OverseasNotifyEmailTemplateConfiguration(
     sourceType: SourceType = SourceType.OVERSEAS,
     receivedEnglish: String,
     receivedWelsh: String,
+    rejectedParentGuardianEnglish: String?,
```
Comment (TomosSlater) [inline src/main/kotlin/uk/gov/dluhc/notificationsapi/config/NotifyClientConfiguration.kt:185]: On the Infra API we've also added "required" comms for the parent guardian proof/qualifying address/documents - not sure if we need them adding here too?

File: src/main/kotlin/uk/gov/dluhc/notificationsapi/database/entity/NotificationType.kt [hunk 3]
```diff
@@ -10,5 +10,7 @@ enum class NotificationType {
     REJECTED_DOCUMENT,
     REJECTED_SIGNATURE,
     REQUESTED_SIGNATURE,
-    NINO_NOT_MATCHED
+    NINO_NOT_MATCHED,
+    REJECTED_PARENT_GUARDIAN,
+    REJECTED_QUALIFYING_ADDRESS
```
Comment (TomosSlater) [inline src/main/kotlin/uk/gov/dluhc/notificationsapi/database/entity/NotificationType.kt:15]: Same comment as above, do we want "REQUIRED_PARENT_GUARDIAN" etc here?

File: src/main/kotlin/uk/gov/dluhc/notificationsapi/dto/NotificationType.kt [hunk 4]
```diff
@@ -29,4 +29,6 @@ enum class NotificationType {
     REQUESTED_SIGNATURE,
     NINO_NOT_MATCHED,
     NINO_NOT_MATCHED_RESTRICTED_DOCUMENTS_LIST,
+    REJECTED_PARENT_GUARDIAN,
```
Comment (TomosSlater) [inline src/main/kotlin/uk/gov/dluhc/notificationsapi/dto/NotificationType.kt:32]: Same comment as above

File: src/main/kotlin/uk/gov/dluhc/notificationsapi/mapper/OverseasDocumentTypeMapper.kt [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+
+    @ValueMapping(source = "PARENT_MINUS_GUARDIAN", target = "PARENT_GUARDIAN")
+    @ValueMapping(source = "QUALIFYING_MINUS_ADDRESS", target = "QUALIFYING_ADDRESS")
+    fun fromApiToDto(overseasDocumentType: OverseasDocumentType): OverseasDocumentTypeDto
+
+    @ValueMapping(source = "IDENTITY", target = "REJECTED_DOCUMENT")
+    @ValueMapping(source = "PARENT_GUARDIAN", target = "REJECTED_PARENT_GUARDIAN")
+    @ValueMapping(source = "QUALIFYING_ADDRESS", target = "REJECTED_QUALIFYING_ADDRESS")
+    fun fromOverseasDocumentTypeDtoToNotificationTypeDto(overseasDocumentType: OverseasDocumentTypeDto): NotificationType
+}
```
Comment (TomosSlater) [inline src/main/kotlin/uk/gov/dluhc/notificationsapi/mapper/OverseasDocumentTypeMapper.kt:20]: Might be a good idea to add a test file for this mapper?

File: src/main/kotlin/uk/gov/dluhc/notificationsapi/mapper/NotificationTypeMapper.kt [hunk 6]
```diff
@@ -49,6 +49,8 @@ interface NotificationTypeMapper {
     @ValueMapping(source = "REJECTED_SIGNATURE", target = "REJECTED_MINUS_SIGNATURE")
     @ValueMapping(source = "REQUESTED_SIGNATURE", target = "REQUESTED_MINUS_SIGNATURE")
     @ValueMapping(source = "NINO_NOT_MATCHED", target = "NINO_MINUS_NOT_MINUS_MATCHED")
+    @ValueMapping(source = "REJECTED_PARENT_GUARDIAN", target = "REJECTED_MINUS_PARENT_MINUS_GUARDIAN")
```
Comment (TomosSlater) [inline src/main/kotlin/uk/gov/dluhc/notificationsapi/mapper/NotificationTypeMapper.kt:52]: We should update the test case in NotificationTypeMapperTest with these new cases

File: src/main/kotlin/uk/gov/dluhc/notificationsapi/config/NotifyClientConfiguration.kt (hunk 2 above)
Comment (ChristianSoftwire) [reply src/main/kotlin/uk/gov/dluhc/notificationsapi/config/NotifyClientConfiguration.kt:185]: Yep that's going to be the next ticket I start

File: src/main/kotlin/uk/gov/dluhc/notificationsapi/config/NotifyClientConfiguration.kt (hunk 1 above)
Comment (ChristianSoftwire) [reply src/main/kotlin/uk/gov/dluhc/notificationsapi/config/NotifyClientConfiguration.kt:224]: Not sure. These templates weren't made by me. This code change was moving the abstract classes to the top of the classes that would inherit from them, just to make things clearer for future readers

File: src/main/kotlin/uk/gov/dluhc/notificationsapi/dto/ContactDetailsDto.kt [hunk 7]
```diff
@@ -15,4 +15,5 @@ data class AddressDto(
     val town: String?,
     val area: String?,
     val postcode: String,
+    val country: String?
```
Comment (wilpow-github) [inline src/main/kotlin/uk/gov/dluhc/notificationsapi/dto/ContactDetailsDto.kt:18]: As discussed this is used for the ERO address and so country is not required

File: src/main/kotlin/uk/gov/dluhc/notificationsapi/mapper/TemplatePersonalisationDtoMapper.kt [hunk 8]
```diff
@@ -176,6 +194,26 @@ class TemplatePersonalisationDtoMapper {
...
+        personalisation["eroWebsite"] = website
+        personalisation["eroEmail"] = email
+        with(address) {
+            personalisation["eroAddressLine1"] = getSafeValue(property)
+            personalisation["eroAddressLine2"] = street
+            personalisation["eroAddressLine3"] = getSafeValue(town)
+            personalisation["eroAddressLine4"] = getSafeValue(area)
+            personalisation["eroAddressLine5"] = getSafeValue(locality)
+            personalisation["eroPostcode"] = postcode
+            personalisation["eroCountry"] = getSafeValue(country)
```
Comment (wilpow-github) [inline src/main/kotlin/uk/gov/dluhc/notificationsapi/mapper/TemplatePersonalisationDtoMapper.kt:213]: Not needed as discussed

File: src/main/kotlin/uk/gov/dluhc/notificationsapi/rest/GlobalExceptionHandler.kt [hunk 9]
```diff
@@ -78,6 +79,7 @@ class GlobalExceptionHandler(
             GovNotifyApiBadRequestException::class,
             GssCodeMismatchException::class,
             NotificationTemplateNotFoundException::class,
+            CountryNotFoundException::class
```
Comment (wilpow-github) [inline src/main/kotlin/uk/gov/dluhc/notificationsapi/rest/GlobalExceptionHandler.kt:82]: Not needed

File: src/main/resources/openapi/NotificationsAPIs.yaml [hunk 10]
```diff
@@ -1411,6 +1549,10 @@ components:
           minLength: 1
           maxLength: 10
           example: PE3 6SB
+        country:
```
Comment (wilpow-github) [inline src/main/resources/openapi/NotificationsAPIs.yaml:1552]: Not needed

File: src/main/resources/openapi/NotificationsAPIs.yaml [hunk 11]
```diff
@@ -1304,6 +1402,46 @@ components:
...
+          items:
+            $ref: '#/components/schemas/RejectedDocument'
+          minItems: 1
+        rejectedDocumentFreeText:
+          type: string
+          description: Free text message to provided further information regarding the document rejection
+          example: Please provide a valid document within 28 days
+      required:
+        - documents
+    BaseOverseasPersonalisation:
```
Comment (wilpow-github) [inline src/main/resources/openapi/NotificationsAPIs.yaml:1423]: Is this any different from BasePersonalisation?

Comment (ChristianSoftwire) [reply src/main/resources/openapi/NotificationsAPIs.yaml:1423]: No I will remove

**Pull request closed without merging**